extensions/amp-form/0.1/amp-form-textarea.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {computedStyle, px, setStyle} from '../../../src/style';
+import {dev, devAssert} from '../../../src/log';
+import {iterateCursor, removeElement} from '../../../src/dom';
+import {listen, listenOncePromise} from '../../../src/event-helper';
+import {throttle} from '../../../src/utils/rate-limit';
+
+const AMP_FORM_TEXTAREA_EXPAND_ATTR = 'autoexpand';
+
+const AMP_FORM_TEXTAREA_SHRINK_DISABLED_ATTR = 'autoshrink-disabled';
+
+const MIN_EVENT_INTERVAL_MS = 100;
+
+const AMP_FORM_TEXTAREA_CLONE_CSS = 'i-amphtml-textarea-clone';
+
+const AMP_FORM_TEXTAREA_MAX_CSS = 'i-amphtml-textarea-max';
+
+const AMP_FORM_TEXTAREA_HAS_EXPANDED_DATA = 'iAmphtmlHasExpanded';
+
+class AmpFormWithTextarea {
+  /**
+   * @param {!Element} form
+   */
+  constructor(form) {
+    /** @private */
+    this.win_ = devAssert(form.ownerDocument.defaultView);
+
+    /** @private */
+    this.unlisteners_ = [];
+
+    this.unlisteners_.push(listen(form, 'input', e => {
```

Comment:
I might've said something like this in chat, but maybe we can reduce the number of event listeners from O(# of forms) to O(1) by:

1. Only installing a listener if there exists a `textarea[autoexpand]` element in the document (and re-query on DOM_UPDATE).
2. Add `input` event listener on the root element instead of each form (similar to action-impl.js).